File: code/0013. Roman to Integer/solution1.java
class Solution {
    public int romanToInt(String s) {
        Map<String, Integer> symbolValues = new HashMap<>();
        symbolValues.put("M", 1000);
        symbolValues.put("CM", 900);
        symbolValues.put("D", 500);
        symbolValues.put("CD", 400);
        symbolValues.put("C", 100);
        symbolValues.put("XC", 90);
        symbolValues.put("L", 50);
        symbolValues.put("XL", 40);
        symbolValues.put("X", 10);
        symbolValues.put("IX", 9);
        symbolValues.put("V", 5);
        symbolValues.put("IV", 4);
        symbolValues.put("I", 1);
        
        int n = s.length();
        int result = 0;
        
        for (int i = 0; i < n; i++) {
            if (i < n - 1 && symbolValues.containsKey(s.substring(i, i + 2))) {
                result += symbolValues.get(s.substring(i, i + 2));
                i++;
            }
            else {
                result += symbolValues.get(s.substring(i, i + 1));
            }
        }
        
        return result;
    }
}

File: code/0013. Roman to Integer/solution1.cpp
class Solution {
public:
    int romanToInt(string s) {
        unordered_map<string, int> symbolValues{
            {"M", 1000}, {"CM", 900}, {"D", 500}, {"CD", 400}, {"C", 100}, {"XC", 90}, 
            {"L", 50}, {"XL", 40}, {"X", 10}, {"IX", 9}, {"V", 5}, {"IV", 4}, {"I", 1}
        };
        
        int n = s.length();
        int result = 0;
        
        for (int i = 0; i < n; i++) {
            if (i < n - 1 && symbolValues.count(s.substr(i, 2)) > 0) {
                result += symbolValues[s.substr(i, 2)];
                i++;
            }
            else {
                result += symbolValues[s.substr(i, 1)];
            }
        }
        
        return result;
    }
};